File: Data Structures & Algorithms/network-delay-time/submission-1.kt
class Solution {
    private data class Target(
        val v: Int,
        val t: Int,
    )
    fun networkDelayTime(times: Array<IntArray>, n: Int, k: Int): Int {
        val x: MutableMap<Int, MutableList<Target>> = mutableMapOf()
        times.forEach { t ->
            if (x.contains(t[0])) {
                x[t[0]]?.add(Target(t[1], t[2]))
            } else {
                x[t[0]] = mutableListOf(Target(t[1], t[2]))
            }
        }
        val tt: MutableList<Int> = MutableList(n){Int.MAX_VALUE}
        val q: MutableList<Target> = mutableListOf(Target(k, 0))
        while (q.isNotEmpty()) {
            val tx: Target = q.removeFirst()
            if (tx.t < tt[tx.v - 1]) {
                tt[tx.v - 1] = tx.t
                if (x.contains(tx.v)) {
                    x[tx.v]?.forEach { tn ->
                        q.add(Target(tn.v, tx.t + tn.t))
                    }
                }
            }
        }
        if (tt.contains(Int.MAX_VALUE)) {
            return -1
        }
        return tt.max()
        
    }
}

File: Data Structures & Algorithms/valid-parenthesis-string/submission-1.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        ub = 0
        lb = 0
        for c in s:
            if c == '(':
                ub += 1
                lb += 1
            elif c == ')':
                if ub == 0:
                    return False
                ub -= 1
                lb -= 1
            else:
                ub += 1
                lb = max(0, lb - 1)
            # print(ub, lb)
        return lb <= 0


        # (**(
